### ISSUE TITLE
Support for other panels

### ISSUE THREAD

**Issue by madmic1314** [RoganDawes/ESPHome_Wintex#1]
Hi there - great work and exactly what I'm looking to do, currently use a PI to do the same, an ESP32 / 8266 would be much better.

Have you seen the work by this chap: https://github.com/dchesterton/texecom2mqtt-hassio he's got a full blown implementation which might help your work along. Again no NDA needed.

Cheers

**Comment by RoganDawes** [comment]
Thanks. I hope it is useful for you.

I had seen that other project, but it requires a Premier Elite panel, not a
simple Premier.

**Comment by madmic1314** [comment]
Ah - I didn't realise there was that much of a difference between the two, I have the Elite 24.

**Comment by RoganDawes** [comment]
Yes, unfortunately. The elite panels support additional protocols that the
non-elite do not. Such as Crestron, etc.

**Comment by madmic1314** [comment]
Does Com-IP not use the serial protocol for both? Shame of not.

**Comment by RoganDawes** [comment]
ComIP actually has some complexity, as there is a device in the ComIP that
the panel can configure IP addresses, etc for. That uses some sort of
AT-based command set implemented in the processor.

My ESPHome approach simply emulates a Com-usb directly connected to the
panel, speaking the same binary protocol that Wintex does.

**Comment by kstormhansen** [comment]
Sorry not familiar with ESP32, but I have a Premier 832 non-elite that I would very much like to integrate into Home Assistant. Would the ComIP also work with the work you have done here? Thanks :-)

**Comment by RoganDawes** [comment]
Well, the concepts still apply, and the protocol hasn’t changed, so that's
good news. The net result, though, is that you would need to implement
something that can poll the alarm panel according to the protocol, and
update hone assistant appropriately.

At some point, I had intentions of implementing this as a custom HA
component, written in Python. Unfortunately, my Python is even worse than
my C/C++ if you can imagine such a thing, so I haven’t got anywhere near
that.

Conceptually, it’s a very simple thing to do: log in to the panel (simple
message with success/fail response), then in a loop, request 16-32 bytes of
memory (depending on how many zones you are interested in), and check one
bit of each byte, which represents the status of the zone. Then update HA’s
data for that zone. It’s the latter part that is most tricky for me, as I
have not looked at direct interfacing with HA in this way.

I don’t know what platforms or programming languages you are familiar with,
but there should be a number of possible ways to get it all working.

Happy to provide more info if you need it.

**Comment by MortenVinding** [comment]
Hi,

I'm a terrible programmer and only really understand bash. But I have made a simple bash script to interface with my Premier 816 panel using only the "simple protocol".

It simply poll's the zones and the log counter, and sends the zone status by MQTT to HA and if the log has advanced it reads the new loglines and reports status changes by MQTT to HA.

it also sends SMS messages if the alarm panel is triggered or there is a loss of 230volt.

I use a small OpenWRT router to run it on, that is powered byt the AUX output of the alarm panel (for battery backup), and also has a 3G USB dongle connected to send SMS messages.

Update times are currently at about 2 secs. because that is how fast I'm able to poll the alarm panel.

It entirely based on reverse engenering the Texecom app, so not using any NDA stuff.

Right now it not in a condition for sharing.
Lots of hardcoded stuff, and reading since it is based on reading the log it is also language dependent, and unfortunately my panel is en Danish (seems there is no way to change that).

But if there is enough intereste, I might be convinced to clean it up and commit it.

Any one, hopefully better programeres, out there who is interested in collaborate on this?
I would love to have it all rewritten in a more appropriate language, preferably Python since that is also well suited for the OpenWRT routers.

**Comment by RoganDawes** [comment]
I'm not really too interested in the Simple protocol, to be honest, even though the Texecom panels do appear to be capable of switching between them fairly easily, even dynamically from one request to the next.

**Comment by MortenVinding** [comment]
> I'm not really too interested in the Simple protocol, to be honest, even though the Texecom panels do appear to be capable of switching between them fairly easily, even dynamically from one request to the next.

fair enough. I was also more thinking if anyone else was interested.
I prefer to leave all the alarm logic to the alarm system and not make it dependent on HA (hence the OpenWRT router, 3G dongle and battery backup), but I'm a bit interested in why you won't use the simple protocol?
Is it too slow or is just not providing the information you need?

I choossed the simple protocol because it was near impossible to find details about the binary protocol and because (mostly) because it looked pretty complex.

**Comment by RoganDawes** [comment]
I guess it is more of an itch than a rational reason not to use the simple protocol. Documentation of the Simple protocol was available, but only under NDA, which I was not interested in. I also got the feeling that the simple protocol was insufficient to achieve all of the desired functionality (although I can't point to something specific that it lacks at this point, I did manage to get hold of the Simple docs without signing an NDA, and I based my decision on that.)
And of course, the Wintex protocol was a completely unknown entity, which became my mountain to climb - because it was there.

Just to note that none of this makes the alarm dependent on HA - it still activates, detects zone intrusions, and alarms exactly as it always has. This simply allows HA to monitor the alarm status and eventually influence some state as though settings were changed using the panel (arming/disarming/bypassing, etc)

**Comment by MortenVinding** [comment]
> I guess it is more of an itch than a rational reason not to use the simple protocol. Documentation of the Simple protocol was available, but only under NDA,
I see. So actually our reasons was the same, just the starting point was different:
You used wintex, and I used the mobile app :)

> 
> Just to note that none of this makes the alarm dependent on HA - it still activates, detects zone intrusions, and alarms exactly as it always has. This simply allows HA to monitor the alarm status and eventually influence some state as though settings were changed using the panel (arming/disarming/bypassing, etc)
Ohh that's interesting. From reading your readme I was under the impression that you code only extracted the zone status.
So actually you get more or less the same info I have (zone status, panel status, set arm/disarm)?

**Comment by RoganDawes** [comment]
Yes, I think our initial departure point was the same, just the route differed.

> Ohh that's interesting. From reading your readme I was under the impression that you code only extracted the zone status.
So actually you get more or less the same info I have (zone status, panel status, set arm/disarm)?

No, you are correct. When I said "it still detects intrusions", etc, I was referring to the alarm panel, not any integration with HA. i.e. all of the normal alarm functions (that work without any connected system) will continue to work.

So the current state of this integration is that it only reports zone status, but it should be relatively simple to add remote arm/disarm.

From the Wintex manual, I see the following panel, which should do the trick. i.e. monitoring the commands that result from selecting the various buttons will tell me how to arm/disarm the panel.

![Screenshot from 2022-03-10 16-12-02](https://user-images.githubusercontent.com/159883/157679663-58d5adbc-3d12-429e-86f0-757643863d5b.png)

**Comment by MortenVinding** [comment]
> Yes, I think our initial departure point was the same, just the route differed.

> No, you are correct. When I said "it still detects intrusions", etc, I was referring to the alarm panel, not any integration with HA. i.e. all of the normal alarm functions (that work without any connected system) will continue to work.
> 
> So the current state of this integration is that it only reports zone status, but it should be relatively simple to add remote arm/disarm.
> 
> From the Wintex manual, I see the following panel, which should do the trick. i.e. monitoring the commands that result from selecting the various buttons will tell me how to arm/disarm the panel.

Thanks.
Well actually maybe our starting points was a bit different, because my goal was to make the alarm usable without a phoneline :)

Sure the report of zone status to Home Assistant is nice, but I needed some way to have it alarm me by SMS and phone calls, and I don't have a phone line anymore.
That's why I made the solution with the OpenWRT router and a 3G dongle.

Thanks for the point on the manual. However I don't think I have the time and the skills to reverse engineer the binary protocol. Especially since my current solution works okay for now.
Sure just hammering on the log like I do now, isen't optimal, but it works. And the update rate is reasonable: about every 2 seconds.

But thank you for your interest. Your solution surely is better, if your panel is already hooked up to a phoneline or even an alarm central.

**Comment by RoganDawes** [comment]
Yeah, my alarm has a radio transmitter to an armed response company, so that part was less important to me. But I do actually want my alarm to notify me directly of the active zone when an alarm is triggered, so I do need to put in a bit more work to detect the activation event.

**Comment by Vivace83** [comment]
@RoganDawes any idea if this work as-is for the Premier 412 panel? I am relatively new to the ESPHome and the Home Assistant space, but was looking to do exactly what u have done here, plus the need to show status(armed/disarmed) and ability to arm and disarm.

**Comment by RoganDawes** [comment]
There's a pretty good chance it will work. I'm busy adding support for Partitions/Areas at the moment, and will then add direct support for arming/stay arming and disarming the panel. Are you able to add an ESP32 to your panel, or do you already have a ComIP or ComWiFi device?

**Comment by Vivace83** [comment]
Going to add the esp32 directly if possible

**Comment by RoganDawes** [comment]
> Going to add the esp32 directly if possible

Sounds good. I used a level converter chip, but you can achieve the same result with a simple voltage divider, between the panel TX and the ESP32 RX pins.

The 3.3V TX from the ESP32 can be connected directly to the panel, because the 3.3v level is sufficient for the panel to read it as a 1.
